### ISSUE TITLE
Emscripten port

### ISSUE THREAD

**Issue by embeddedt** [lvgl/lvgl#687]
I've completed an Emscripten port. Would you be interested in adding this to the `lv_projects` repository?

![screenshot from 2019-01-01 17-27-42](https://user-images.githubusercontent.com/42941056/50576937-b1eb3180-0dea-11e9-8882-653259c02683.png)

**Comment by amirgon** [comment]
@embeddedt Does it include Micropython?

**Comment by embeddedt** [comment]
@amirgon At the moment, no. It is just a standard port of LittlevGL.

**Comment by amirgon** [comment]
> At the moment, no. It is just a standard port of LittlevGL.

If the purpose is to let the user experience lvgl on a web page, I think that Micropython is essential. Think of a web page where the user can interactively run lvgl online through a web based Micropython REPL.

**Comment by embeddedt** [comment]
@amirgon I will give it a try!

Is there any official MicroPython port for Emscripten? A quick Google search is showing:

* micropython/micropython#3575 (pure JavaScript)
* https://github.com/micropython/micropython-unicorn (running on top of Unicorn.js)
* https://github.com/numworks/epsilon/tree/master/python.

**Comment by amirgon** [comment]
> @amirgon I will give it a try!

:+1: 

I would say Unicorn is the most "official" one. It is hosted under Micropython URL (https://micropython.org/unicorn/) so maybe this is the one to start with, at least until https://github.com/micropython/micropython/pull/3575 is accepted into Micropython.

**Comment by embeddedt** [comment]
@amirgon Okay, I will use that one.

Even after I finish the MP port I think that having regular Emscripten support would be good as well.

**Comment by amirgon** [comment]
> Even after I finish the MP port I think that having regular Emscripten support would be good as well.

@embeddedt out of curiosity, what is the use case of the "regular" Emscripten support?  
If I just want to create GUI on a web page there are many other more conventional ways to do it, utilizing HTML, CSS and JavaScript.  
I can see how Emscripten support is useful for lvgl showcase, or even some "development environment" to design GUI online with Micropython. I just don't understand the other use cases.

**Comment by kisvegabor** [comment]
@embeddedt 
> I've completed an Emscripten port. Would you be interested in adding this to the lv_projects repository?

Awesome! Yes, please, add it to `lv_projects` even without Micropython support. I'm very curious about it.

@amirgon 
> @embeddedt out of curiosity, what is the use case of the "regular" Emscripten support?

I can't imagine a better way to add a running GUI to webpage. With this we can add some GUIs to littelvgl.com. The user can see what kind of objects, styles and animations are there. 

However, with Micropython support, it would be an interactive GUI editor which works without installation on every platform. It'd be amazing!

**Comment by embeddedt** [comment]
@amirgon 

> @embeddedt out of curiosity, what is the use case of the "regular" Emscripten support?

As @kisvegabor said, having a live GUI running on the webpage would be a lot more immersive than screenshots.

**Comment by embeddedt** [comment]
@kisvegabor Please see the `emscripten` directory inside `lv_projects`. After initializing submodules, as long as the Emscripten SDK is in your path you should be able to run `make` and everything should build.

**Comment by kisvegabor** [comment]
@embeddedt I was thinking about a new menu item called "Live demo".

I'd be great to replace the opening image too. But I'm afraid of the page load time. Maybe a normal jpg image can be displayed first and when the demo is loaded we can replace the image.

**Comment by embeddedt** [comment]
@kisvegabor 

> Maybe a normal jpg image can be displayed first and when the demo is loaded we can replace the image.

I like that approach.

**Comment by kisvegabor** [comment]
@embeddedt It also crashes in `demo` app on the Lis tab when clicking "Got it". It also closes a message box. Maybe it's an issue in the library.

**Comment by kisvegabor** [comment]
I've tried with `lv_test_stress_1();` and it crashed immediately. I suspect that the object deletion or memory reallocation causes the issue.

**Comment by kisvegabor** [comment]
I've played with the Debug options if `emcc` to see stack trace or similar but with no luck so far... :(

**Comment by embeddedt** [comment]
@kisvegabor I pushed the debug-enabled Makefile to `master`. Does that help?

**Comment by kisvegabor** [comment]
@embeddedt 
Thank you for the Makefile.

I've fixed the issue! It was a very tricky one. The crash was in `anim_ready_handler` because of this line:
`void (*cb)(void *) = a->end_cb;`

When a message box closes it plays a shrinking animation with `lv_obj_del` `end_cb`. But `lv_obj_del` has `lv_res_t` return value instead of `void`. And, I don't know why, but this mismatch in the return values caused the crash. I've replaced `lv_obj_del` with this function to have the correct return value and now it works:
```c
static void lv_mbox_close_end_cb(lv_obj_t * mbox)
{
    lv_obj_del(mbox);
}
```

Do you have any idea why did it solve the issue?

**Comment by embeddedt** [comment]
@kisvegabor I'm really happy that you were able to fix it!

> Do you have any idea why did it solve the issue?

On all architectures I have used you can get away with using `lv_obj_del` as if it returned `void`. However, it wouldn't surprise me to learn that the way Emscripten handles return values (especially because it is a transpilation to JavaScript), this kind of trickery is not allowed. Do you know if this is considered undefined behavior in C? If so, I'm concerned that these tricks might be used in other places.

**Comment by kisvegabor** [comment]
I've never experienced any issues with ignoring the return value. But yes, it might be something specific to Emscripten. Anyway, if we find something similar with Emscripten, we should remember this. :)

**Comment by kisvegabor** [comment]
I think it's ready to add some live demos to the website.  Do you know a good way to remove the unused parts? Such as Emscripteh header, console etc.

**Comment by embeddedt** [comment]
@kisvegabor I'll look into that now, and I will adjust the Makefile if and when I figure it out.

**Comment by kisvegabor** [comment]
Great!

**Comment by embeddedt** [comment]
@kisvegabor It's been pushed to `master`.

**Comment by kisvegabor** [comment]
Awesome! Exactly what I imagined. It would take hours for me to do it. :)

**Comment by kisvegabor** [comment]
I will add some examples to the website next week.

**Comment by embeddedt** [comment]
@kisvegabor I was really interested in discovering exactly why the return value was a problem and came across kripken/emscripten#2872.

Turns out that this is undefined behavior, but Emscripten can allow this behavior if you pass `-s EMULATE_FUNCTION_POINTER_CASTS=1` when compiling. I added a note in the Makefile explaining this.

**Comment by embeddedt** [comment]
@kisvegabor It runs through the stress test with no issues now.

**Comment by kisvegabor** [comment]
So the previous version of the message box is working with EMULATE_FUNCTION_POINTER_CASTS?

**Comment by embeddedt** [comment]
@kisvegabor Yes. If `EMULATE_FUNCTION_POINTER_CASTS` is used then your extra function is not necessary. However, I think we should leave that option off for now because it can limit browser compatibility and performance (although I haven't had any issues thus far).

How much code does the stress test test? I think we have to make sure that this kind of casting isn't used elsewhere.

**Comment by embeddedt** [comment]
@kisvegabor In that case I suppose we can consider the Emscripten port stable. At least for now.

**Comment by kisvegabor** [comment]
@embeddedt I agree!

**Comment by kisvegabor** [comment]
I've created an emscripten repo. Can you move the project there? https://github.com/littlevgl/emscripten

**Comment by embeddedt** [comment]
@kisvegabor Project has been uploaded there.

**Comment by kisvegabor** [comment]
Thank you!

**Comment by amirgon** [comment]
Very nice. Is there a way to make it scale better on mobile? It works but screen is truncated.

**Comment by kisvegabor** [comment]
I've searched for it and there are some JavaScript-based solutions.
I didn't try it but I'm afraid the mouse position might be incorrect when scaled.

**Comment by kisvegabor** [comment]
@embeddedt 
As the Emscripten port is ready and Live demos are running, I think we can close this issue.

**Comment by amirgon** [comment]
@embeddedt @kisvegabor Did you consider running Micropython in the Emscripten port, as I discussed [above](https://github.com/littlevgl/lvgl/issues/687#issuecomment-450866404)?  
Could be useful as an interactive sandbox to let the users experience lvgl on the browser without compiling or installing anything.

**Comment by puzrin** [comment]
@amirgon 

> Could be useful as an interactive sandbox to let the users experience lvgl on the browser without compiling or installing anything.

Such technical concept is not new and used wide, but success is absolutely random. I would not take responsibility to recommend you start spending your time immediately.

First, need to make one step back and decide, why such sandbox could be VERY useful at all (help to solve everyday user's tasks with ease).

- "playing in sandbox" is too abstract for "all users", not task from real world.
- sharing live snippets (via gist, for example) may be interesting, but why all participant could be motivated?
  - Why snippet author can be interested to spend time for create snippet?
  - Will user be able to use snipet in his project as is or with minimal change?

Minimal requirement for success - describe one of more particular workflows (step-by step), where users are interested to participate in each step and happy in the end. That's not absolute guarantee, but flag "can this be successive at all".

Note, i did not mentioned any specific language above. All such questions should be applied to C, embedded python, embedded js, embedded lua. Using C online is not principal limitation, because you can compile sources on server side (see https://mbebenita.github.io/WasmExplorer/).

For example, if i do projects in C, why can i be interested to write demo in python? If i found demo in python, can i transpile it in one click to C (how can i use it with ease for my needs)?

On other words, "demo interactivity" is not thing users need directly. They need easy knowledge share. If you can explain, why such share is awesome via interactive python demo - then it worth to try.

**Comment by amirgon** [comment]
> First, need to make one step back and decide, why such sandbox could be VERY useful at all (help to solve everyday user's tasks with ease).

I can think of some use cases for a sandbox:

- User wants to have a quick look at some lvgl object, see how it looks like and whether it fits his needs. You can look at the docs, but they won't cover every aspect of every object and will not reflect everything visually. Especially if you want to see the relation of some objects together, styles, themes etc.
- User wants to report a bug or a problem. A short python script could allow anyone reproduce it immediately in the browser and is easily shareable.
- User wants to quickly evaluate lvgl for some project. One option is to download some project  and play with it, but interactive sandbox is much more immediate and accessible.
- User finds lvgl when browsing the web. "Screenshots look nice but what can I actually do with that?" Now imagine a web text area with some python code that the user can change and immediately see the result. A much better experience than requiring the user to install and compile stuff before getting the first "real" impression of lvgl.

> Note, i did not mentioned any specific language above

Yes you could do that with C as well, if you want.
In my opinion Micropython would be simpler, shorter and more accessible for a larger audience than C, but same arguments apply to any other language that has "lvgl binding" including C.

**Comment by puzrin** [comment]
> In my opinion Micropython would be simpler, shorter and more accessible for a larger audience than C, but same arguments apply to any other language that has "lvgl binding" including C.

If i understand you right, this is about collecting new community from zero "because python is cool". Start from zero is very risky.

You could have more chances, if you already know existing MP community and know how to motivate them use lvgl.

> but same arguments apply to any other language that has "lvgl binding" including C.

I understand arguments about scripting language simplicity (there are some details to discuss, but general idea is correct). But we can not have expectation that users _already know_ python.

So, i disagree about "any" language due simple math: number of "abstract" js users is > of python users. That doesn't mean js is better. But your chances to form stable community from zero is higher.

I don't mean anything against python. Only try to explain, how to estimate chances of success, prior to start.

**Comment by embeddedt** [comment]
@puzrin @amirgon

> You could have more chances, if you already know existing MP community and know how to motivate them use lvgl.

[Quote:](https://github.com/littlevgl/lvgl/issues/557#issuecomment-457888980)

> The MicroPython community really needs a decent UI library. LittlevGL is certainly an excellent option and I'm going to watch the integration carefully and help out where I can.

I would say that the motivation is present. When I have read about GUIs on MicroPython in the past there is really no official GUI library that everyone uses. I see this as an excellent opportunity for LittlevGL.

However, I agree that using MicroPython as a sandbox might not be the best option. It would really only be useful to those who are already familiar with Python's syntax, or can grasp it quickly.

**Comment by puzrin** [comment]
In other words, such sandbox will be not "lvgl feature", but "micropython feature".

That's not bad, but will not intersect with lvgl C users anyhow.

**Comment by kisvegabor** [comment]
Form community building's point of view the goal of a sandbox like this should be to ensure that people will have a happy first experience with LittlevGL. Instead of 
- IDE/makefile
- SDL or other drivers
- git clone, submodules
- build etc

If people already tried and liked lvgl, it's more likely that they will take the effort to set up a real development environment. It's good for SEO too because more time spent on a site means a better ranking. (I know it's oversimplified but in general true).

True, only a group of people will use LittlevGL with Micropython, but:
1. Python is known by a lot of people.
![python popularity](https://cdn-images-1.medium.com/max/1600/0*5sXl34xnPk7LW6YP)
[source](https://hackernoon.com/top-3-most-popular-programming-languages-in-2018-and-their-annual-salaries-51b4a7354e06)
Actually, I can't imagine a better script language binding for LittlevGL. 
2. We can add very simple examples which can be understood by somebody who actually doesn't know Python. E.g.  it would be clear to any programmer and can play with it
```python
>>> import lvgl
>>> btn = lvgl.btn(lvgl.scr_act())
>>> btn.align(lvgl.scr_act(), lvgl.ALIGN.CENTER, 0, 0)
>>> label = lvgl.label(btn)
>>> label.set_text("Button")
```
3. Altoght Python is very different from C  it can be used "C-ish" way.  So don't write 
```python
lvgl.btn(lvgl.scr_act()).align(lvgl.scr_act(), lvgl.ALIGN.CENTER, 0, 0)`
```
4. People still will know how LittlevGL works in principle and can decide if it is suitable for them or not. 
6. And finally, it seems it's the most simple to realize. A C compiler on a server which compiles the sent code and sends back the ready image (or errors) is less user-friendly, slower and requires a server. Running it on the client side is more scalable. 

So in summary, I think a Micropython - Emscripten port would be very useful!

**Comment by embeddedt** [comment]
@kisvegabor 

> So in summary, I think a Micropython - Emscripten port would be very useful!

We determined before that using Micropython on top of Unicorn.js (an emulator) was the most supported approach. So it wouldn't be a direct port, but more of an emulation. Is that alright with you?

**Comment by kisvegabor** [comment]
@embeddedt 
> We determined before that using Micropython on top of Unicorn.js (an emulator) was the most supported approach. So it wouldn't be a direct port, but more of an emulation. Is that alright with you?

If you think Unicorn.js is the best then I'm absolutely happy with it. Knowing that it's hosted by micropython sounds very well. I don't know how to put together lvgl-emscripten and mpy-unicorn.js but I suppose you have the idea.

**Comment by amirgon** [comment]
@kisvegabor @embeddedt   
Unicorn.js is definitely an option. It has, however, some disadvantages:
 
- It may be too slow. It's an emulation of an ARM processor running on the browser, so it's kind of an emulation (of C code on the browser) of an emulation (of ARM core in C code).  
My guess is that animations are out of the question, and I'm not sure how responsive it can get, but maybe it's worth a try.  
https://micropython.org/unicorn/ has some demo examples that show how to use an i2c display. I tried them and wasn't impressed at all by the speed and responsiveness.
- It will require developing input device driver and display driver for that ARM architectures that interface with the web browser. It's possible but probably requires some extra effort, because the actual architecture hardware needs to be used (i2c, spi, DMA etc.)

**There is another option**: running Micropython+lvgl directly on the browser using Emscripten.  
It will definitely run faster than Unicorn.js (no need to emulate ARM and all the hardware peripherals), might even run smooth animations, since WebASM / ASM.JS are not bad at running pure C code.  
It will require display and input drivers but I suspect they would be simpler than the Unicorn.js option because we could "bypass" the "real" hardware emulation (i2c, spi, DMA etc.) and access the canvas on the web page directly.

**Comment by embeddedt** [comment]
@amirgon 

I'm not pointing fingers here, but earlier I asked you about which Micropython port you would recommend, and you said:
> 
> +1
> 
> I would say Unicorn is the most "official" one. It is hosted under Micropython URL (https://micropython.org/unicorn/) so maybe this is the one to start with, at least until [micropython/micropython#3575](https://github.com/micropython/micropython/pull/3575) is accepted into Micropython.

Have there been any new developments in Micropython+Emscripten support since then?

**Comment by amirgon** [comment]
> I'm not pointing fingers here, but earlier I asked you about which Micropython port you would recommend, and you said:

@embeddedt You are right, I really thought it was the best option.  
Then after the discussion above I started thinking :thinking: ... and started to have some doubts about unicorn.js. Then I played a bit with the display module on https://micropython.org/unicorn/ and tried to get some impression about the speed and responsiveness. Now my machine is an old Dell Inspiron laptop, but still, it just felt too slow.  
Unicorn.js is still a valid option, I just felt it would be a good idea to let you know my thoughts.
So the only new development is me changing my mind :wink:

**Comment by embeddedt** [comment]
@kisvegabor Emscripten can't perfectly compile all C code, and I have read that there are issues compiling Micropython to Emscripten directly.

**Comment by amirgon** [comment]
@embeddedt Maybe it's worth trying to build micropython/micropython#3575? On that PR some people reported success, maybe we'd also be lucky.

**Comment by embeddedt** [comment]
@amirgon I'll give that a try in a few hours.
